### ISSUE TITLE
Tuning on Transformer RNN-T

### ISSUE THREAD

**Issue by zrsjta** [espnet/espnet#2606]
Hi! Thanks for this great project! :)
I am trying to tune a Transformer RNN-T on Aishell-1 dataset, following your config in `egs/vivos/asr1/conf/tuning/transducer/train_transformer_transducer.yaml` with some modification. I have extended the model into a much larger one, but encountered a over-fitting problem and a serious performance degradation compared with the LSTM version. Below is the details.

python:3.7.7
pytorch: 1.0.0
CUDA: 10.1
RNN-T binding: warp-transducer

All experiments are done with 4 Tesla M40 GPUs 

A typical config I use: 

```
batch-size: 30
maxlen-in: 800
maxlen-out: 150

criterion: loss
early-stop-criterion: "validation/main/loss"
sortagrad: 0
opt: noam
epochs: 40
patience: 0
accum-grad: 1
grad-clip: 5.0

transformer-lr: 1.0  # use lower lr for larger transformer
transformer-warmup-steps: 8000

etype: transformer
transformer-enc-input-layer: vgg2l
enc-block-arch:
        - type: transformer
          d_hidden: 768
          d_ff: 768
          heads: 12
          att-dropout-rate: 0.4
enc-block-repeat: 12

dtype: transformer
transformer-dec-input-layer: embed
dec-block-arch:
        - type: transformer
          d_hidden: 768
          d_ff: 768
          heads: 12
          dropout-rate: 0.1
          pos-dropout-rate: 0.1
          att-dropout-rate: 0.1
dec-block-repeat: 6

joint-dim: 768

rnnt-mode: 'rnnt' 
model-module: "espnet.nets.pytorch_backend.e2e_asr_transducer:E2E"
```

（1）When I use a LSTM version of RNN-T, with encoder "vgglstmp", 3 layer, 1024 hidden, and decoder "lstm", 1 layer, 1024 hidden, the test WERs are 7.8%/8.9% for dev/test set. with final training loss around 0.15 and valid loss around 5.13
（2）When I use the transformer version of RNN-T, with the config above, though the training loss could reach the same level(around 0.15), the valid loss is around 7, which is much worse than the LSTM version. Also, the valid loss would grow higher or fluctuate even during the first half of training without getting lower any further. So I assume this is an over-fitting problem 
（3）I find SpecAugment is much helpful to alleviate this problem: similar valid loss with that of LSTM version. But I want to bypass the use of that trick.
（4） I have also tried to: A. use a much smaller encoder-decoder layer number and hidden size (8 encoder, 4 decoder and 512 hidden); B. change the att-dropout-rate of encoder from 0.4 to 0.1; C. Change the d_ff size to 2048; D. tuning the learning rate. All of these attempts leads to similar or even worse results on dev_loss. 
（5）For those with poor valid_loss, I have not do the decoding stage to check the final WER but those CERs can hardly be better than that of LSTM version, I guess. I would do the decoding and report here later if needed.
（6）All experiments are done with 4 GPUs. the total batch size is 120 so I disable the accum-grad setting.

It seems that the config above consumes around 100M parameter budget but is still poor in performance. How should I tune a good config? Thanks for anyone who can help!

regards!

**Comment by b-flo** [comment]
Hi,

> （1）When I use a LSTM version of RNN-T, with encoder "vgglstmp", 3 layer, 1024 hidden, and decoder "lstm", 1 layer, 1024 hidden, the test WERs are 7.8%/8.9% for dev/test set. with final training loss around 0.15 and valid loss around 5.13

I'm not too familiar with AISHELL-1 but I assume it's CER instead of WER. DId you use a LM during decoding?

> （2）When I use the transformer version of RNN-T, with the config above, though the training loss could reach the same level(around 0.15), the valid loss is around 7, which is much worse than the LSTM version. Also, the valid loss would grow higher or fluctuate even during the first half of training without getting lower any further. So I assume this is an over-fitting problem
> （4） I have also tried to: A. use a much smaller encoder-decoder layer number and hidden size (8 encoder, 4 decoder and 512 hidden); B. change the att-dropout-rate of encoder from 0.4 to 0.1; C. Change the d_ff size to 2048; D. tuning the learning rate. All of these attempts leads to similar or even worse results on dev_loss.
>（5）For those with poor valid_loss, I have not do the decoding stage to check the final WER but those CERs can hardly be better than that of LSTM version, I guess. I would do the decoding and report here later if needed.

Some quick notes:

1) Don't directly compare loss values. It's a training indicator, yes, but it doesn't directly translate to performance. You can have a lower CER/WER and higher loss with tranformer-based models compared to RNN-based ones. I would suggest monitoring CER/WER but it may significantly impact training speed in ESPnet1.
2) You can also try `opt: adadelta` with `criterion: loss_eps_decay_only`, it was found useful in some T-T experiments and it may be easier if you only monitor loss.
2) Checkpoint averaging is normally used for transformer/conformer decoding but it's not enabled by default for transducer-based models (because of the customizable architecture). To use it, add the following lines to the condition at l.231 in AISHELL-1's `run.sh`:
    ```
    ...
    [[ $(get_yaml.py ${train_config} etype) = *transformer* ]] || \
    [[ $(get_yaml.py ${train_config} dtype) = *transformer* ]]; then
    ```
    @sw005320 On that note, should I modify all recipe running scripts?
3) On some dataset, I found transformer-transducer way harder to train than RNN-transducer. Also, the former won't always beat the later. You may try to replicate architecture and training parameters from https://arxiv.org/pdf/1909.13037.pdf, it's trained on AISHELL-1.
4) If you're trying to find the best model possible, go for conformer blocks instead of transformer. Also, while it was found out conformer is more sensitive to some parameters such as learning rate compared to transformer, I observed the opposite for transducer.
5) You can stick to 1/2-layer LSTM for decoder, maybe increase number of layers later if needed.
6) If you observe a diverged training, you could augment number of warm-up steps and maybe gradient clipping.
8) Without extra informations (training curves, att plot, etc), I can really say what's the issue honestly. Just by looking at the config you shared and my experiments with similar dataset, I would suggest:
    - reducing `heads` (12 -> 4/8)
    - reducing `d_hidden` (768 -> 256/512)
    - augmenting `d_ff` (768 -> 1024/2048)
    - reducing number of blocks (12 -> 8 for enc., 6 -> 2/4 for dec. or use LSTM as I previously said)
    - reducing `joint-dim` (768 -> 512)
    - reducing encoder `att-dropout-rate` (0.4 -> 0.2) and augmenting `dropout-rate` and `pos-dropout-rate` (0.0 -> 0.1/0.2).

**Comment by sw005320** [comment]
> Checkpoint averaging is normally used for transformer/conformer decoding but it's not enabled by default for transducer-based models (because of the customizable architecture). To use it, add the following lines to the condition at l.231 in AISHELL-1's run.sh:
...
[[ $(get_yaml.py ${train_config} etype) = *transformer* ]] || \
[[ $(get_yaml.py ${train_config} dtype) = *transformer* ]]; then
@sw005320 On that note, should I modify all recipe running scripts?

Yes, let's do that.

I think the number looks good as an initial trial.
If we have better numbers, I'd like to ask @zrsjta to share the config.
This would be very beneficial for other users.

**Comment by zrsjta** [comment]
Hi! Thanks for your advice!  @b-flo 

All of my experiments are done with default LM rescore setting in the recipe. Sorry for not mention that early.

Following your advice, I have tried the Transformer-Encoder, LSTM decoder setting as below
`
batch-size: 30
maxlen-in: 800
maxlen-out: 150

criterion: "loss_eps_decay_only"
early-stop-criterion: "validation/main/loss"
sortagrad: 0
opt: noam
epochs: 30
patience: 0
accum-grad: 1
grad-clip: 5.0

transformer-lr: 1.0
transformer-warmup-steps: 8000

etype: transformer
transformer-enc-input-layer: vgg2l
enc-block-arch:
        - type: transformer
          d_hidden: 512
          d_ff: 2048
          heads: 4
          att-dropout-rate: 0.1
          dropout-rate: 0.1
          pos-dropout-rate: 0.1
enc-block-repeat: 8

dtype: lstm
dlayers: 1
dec-embed-dim: 512
dunits: 512
dropout-rate-decoder: 0.1
dropout-rate-embed-decoder: 0.1

joint-dim: 512

rnnt-mode: 'rnnt' # switch to 'rnnt-att' to use transducer with attention
model-module: "espnet.nets.pytorch_backend.e2e_asr_transducer:E2E"
`
The results are as follow, and the training log is in the bottom
`
write a CER (or TER) result in ../exp/train_sp_rnn-transformer-small/decode_dev_decode_default_lm/result.txt
  | SPKR    | # Snt   # Wrd  | Corr     Sub     Del     Ins     Err   S.Err  |
  | Sum/Avg |14326   205341  | 91.0     8.2     0.8     0.3     9.2    53.3  |
`
`
write a CER (or TER) result in ../exp/train_sp_rnn-transformer-small/decode_test_decode_default_lm/result.txt
 | SPKR    | # Snt   # Wrd  | Corr      Sub     Del     Ins     Err   S.Err  |
 | Sum/Avg | 7176   104765  | 90.1      8.9     1.0     0.3    10.2    56.1  |
`
As a reference, the dev/test result for a LSTM encoder-decoder architecture achieve 92.4%/91.3% (3 layer encoder, 1 layer decoder, 1024 hidden), which is better than this.
[log.txt](https://github.com/espnet/espnet/files/5463493/log.txt)

Here are some questions:
(1) In the training los, the dev_loss increment in last epochs, why would that happen and how to solve it?
(2) I'm not very clear that, though the default recipe for a ctc-attention setting use a very large network(18 layer transformer with 2048 hidden size) , you recommend to use a much smaller one. As we have discussed before, a much large model would not lead to a better performance, so how should we tune this further?  Is there any insight behind this? 

Thanks

**Comment by b-flo** [comment]
> (1) In the training los, the dev_loss increment in last epochs, why would that happen and how to solve it?

Overfitting would be my first guess but as I said loss and accuracy are not exactly correlated so I would also monitor CER/WER, training-validation gap and other things before drawing conclusion. 
On how to solve it, it depends but general advices would be to increase dropout rate or add some regularization, try different initialization, etc.

> (2) I'm not very clear that, though the default recipe for a ctc-attention setting use a very large network(18 layer transformer with 2048 hidden size) , you recommend to use a much smaller one. As we have discussed before, a much large model would not lead to a better performance, so how should we tune this further? Is there any insight behind this?

Sorry if it was misleading. I didn't say that "a much large model would not lead to a better performance",  I only suggested reducing some parameters and augmenting other ones based on the config you provided and similar experiments only.
As I said, I'm not too familiar with AISHELL-1 so if I had to train a model on it myself I would first try to replicate the model and training from the paper I previously shared (https://arxiv.org/pdf/1909.13037.pdf) or try to use the architecture from the default ctc-att config with transducer.

Also, just to be clear on the terminology between CTC-Att and Transducer architecture: `(e|d)units` and `adim` in the first one corresponds to `d_ff` and `d_hidden` in the latter.

P.S: I don't have any resources to allocate right now but I'm planning to make transducer configs for most recipe in a near future, I added Aishell-1 to my list.

**Comment by Peidong-Wang** [comment]
Is there anything I can help on transducer configs? @b-flo Most companies use transducer decoders (at least for first-pass recognition) but it has been missing in ESPnet.

**Comment by b-flo** [comment]
@Peidong-Wang Yeah I know, it's often requested but unfortunately I'm lacking ressources for that and I prefer to allocate them on new/necessary features.
I'm currently "on break" for personal reasons, I'll resume and finish current items and I'll start working on recipes after that. Any help is appreciated, I'll open a new issue at that time!

**Comment by lyqlola** [comment]
@zrsjta could you please share your config for the lstm version?
> （1）When I use a LSTM version of RNN-T, with encoder "vgglstmp", 3 layer, 1024 hidden, and decoder "lstm", 1 layer, 1024 hidden, the test WERs are 7.8%/8.9% for dev/test set. with final training loss around 0.15 and valid loss around 5.13

**Comment by okkteam** [comment]
> （1）When I use a LSTM version of RNN-T, with encoder "vgglstmp", 3 layer, 1024 hidden, and decoder "lstm", 1 layer, 1024 hidden, the test WERs are 7.8%/8.9% for dev/test set. with final training loss around 0.15 and valid loss around 5.13

@zrsjta Hi,
I want to know whether the 7.8%/8.9% CER obtained on aishell-1 dev/test set is trained by `vgglstmp` encoder instead of `vggblstmp`? 

I have tuned an RNN-T model with encoder `vgglstm`, 8 layers, 1024 hidden, and decoder `lstm`, 2 layers, 1024 hidden. The final train/valid loss around 0.7/8.23 at epoch 18, and got much worse CER.

**Comment by b-flo** [comment]
FYI: It seems using multi GPUs results in some degradation from my recent experiments. I usually train with a single GPU so I didn't pay attention but with all the changes we may have introduced a bug somewhere impacting that. I'll investigate.

> I have tuned an RNN-T model with encoder vgglstm, 8 layers, 1024 hidden, and decoder lstm, 2 layers, 1024 hidden. The final train/valid loss around 0.7/8.23 at epoch 18, and got much worse CER.

What about performance with the parameters he mentioned? Also, did you modify `accum-grad` according to the number of GPUs used?

**Comment by okkteam** [comment]
> What about performance with the parameters he mentioned? Also, did you modify `accum-grad` according to the number of GPUs used?

I use 1 GPU and `accum-grad=1`. 
He used the uni-directional RNN Transducer model and got an 8.9% CER on Aishell-1 test set. According to my fine-tuning experience, I only get nearly result on the bi-directional RNN-T model with `4-layer 512-unit blstmp` encoder ,`1-layer 512-unit lstm` decoder and employed speed-perturbed.

**Comment by okkteam** [comment]
> What about performance with the parameters he mentioned? Also, did you modify `accum-grad` according to the number of GPUs used?

I use 1 GPU and `accum-grad=1`. 
He used the uni-directional RNN Transducer model and got an 8.9% CER on Aishell-1 test set. According to my fine-tuning experience, I only get nearly result on the bi-directional RNN-T model with `4-layer 512-unit blstmp` encoder ,`1-layer 512-unit lstm` decoder and employed speed-perturbed.

**Comment by okkteam** [comment]
> OK thank you! After current PRs are donne, I'll allocate some time on a transducer recipe for aishell as it seems needed. It may be a good candidate to finetune current model and pending training features/modifications.

It's my pleasure to help!
